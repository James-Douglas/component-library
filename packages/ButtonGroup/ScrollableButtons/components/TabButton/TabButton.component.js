import React, { useRef, useEffect } from 'react';
import PropTypes from 'prop-types';

import { Link } from '@comparethemarketau/manor-link';
import { TabButtonContainer, TextContainer } from './TabButton.styles';

const TabButton = ({
  href, label, onClick, selected, onRef, handleOnClick,
}) => {
  const ref = useRef(null);

  useEffect(() => {
    if (onRef) {
      onRef(ref);
    }
  }, []);

  const handleClick = () => {
    if (onClick && !selected) {
      onClick();
      handleOnClick();
    }
  };

  return (
    <TabButtonContainer ref={ref} selected={selected}>
      <Link trackingLabel={label} href={href} underline="none" handleClick={handleClick}>
        <TextContainer selected={selected}>
          {label}
        </TextContainer>
      </Link>
    </TabButtonContainer>
  );
};

TabButton.propTypes = {
  selected: PropTypes.bool.isRequired,
  href: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  onClick: PropTypes.func,
  onRef: PropTypes.func,
  handleOnClick: PropTypes.func,
};

TabButton.defaultProps = {
  onClick: () => { },
  onRef: () => { },
  handleOnClick: () => { },
};

export default TabButton;
