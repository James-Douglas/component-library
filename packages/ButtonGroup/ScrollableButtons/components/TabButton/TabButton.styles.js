import styled from 'styled-components';
import { Typography } from '@comparethemarketau/manor-typography';

export const TabButtonContainer = styled.li`
  display: inline-block;
  list-style: none;
  ${({ selected, theme }) => selected && `
    p {
      border-bottom: 0.2rem solid ${theme.colors.primary500};
    }
  `}

  a:focus-visible {
    outline: ${({ theme }) => theme.colors.primary500} auto 1px;
    outline-offset: -1px;
  }

  /*
  * @media (hover: hover): Prevent sticky hover effects for buttons on touch devices
  * If the browser supports proper/true/real/non-emulated hovering, then apply styles included when the button is hovered over.
  * ref:1: https://www.w3.org/TR/mediaqueries-4/#descdef-media-hover
  *     2: https://stackoverflow.com/questions/17233804/how-to-prevent-sticky-hover-effects-for-buttons-on-touch-devices
  */
  @media (hover: hover) {
    :hover {
      background: ${({ theme }) => theme.button.tertiary.backgroundHover};
  }
`;

export const TextContainer = styled(Typography)`
  padding-top: ${({ theme }) => theme.spacing['12']};
  padding-right: ${({ theme }) => theme.spacing['24']};
  padding-bottom: ${({ theme }) => theme.spacing['12']};
  padding-left: ${({ theme }) => theme.spacing['24']};
  font-size: ${({ theme }) => theme.fontSize.sm};
  color: ${({ theme }) => theme.colors.grey700};

  ${({ selected, theme }) => selected && `color: ${theme.colors.primary500};
      &:after {
        font-size: ${theme.fontWeight.bold};
      }`}
`;
