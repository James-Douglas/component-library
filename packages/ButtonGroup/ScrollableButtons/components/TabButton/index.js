export { default } from './TabButton.component';
