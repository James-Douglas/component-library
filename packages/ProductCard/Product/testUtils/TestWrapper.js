import React from 'react';
import ProductProvider from '../../Context/ProductProvider';

const ProductTestWrapper = ({
  // eslint-disable-next-line react/prop-types
  children, product, onDetails = () => {}, onSelect = () => {},
}) => (
  <ProductProvider product={product} onDetails={onDetails} onSelect={onSelect}>
    { children }
  </ProductProvider>
);

export default ProductTestWrapper;
