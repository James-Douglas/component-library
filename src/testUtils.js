// test-utils.js
import {
  render, fireEvent, wait, act,
} from '@testing-library/react';
import ManorProvider from './providers/ManorProvider';

const customRender = (ui, options) => render(ui, { wrapper: ManorProvider, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export {
  customRender as render,
  // export other used functions to appease eslint
  fireEvent,
  wait,
  act,
};
