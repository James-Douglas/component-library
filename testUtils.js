import React from 'react';
import {
  render, fireEvent, wait, act,
} from '@testing-library/react';
import ManorProvider from './packages/Provider/ManorProvider';

const ManorTestWrapper = ({ children }) => (
  <ManorProvider>
    {children}
  </ManorProvider>
);

const customRender = (ui, options) => render(ui, { wrapper: ManorTestWrapper, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export {
  customRender as render,
  // export other used functions to appease eslint
  fireEvent,
  wait,
  act,
};
