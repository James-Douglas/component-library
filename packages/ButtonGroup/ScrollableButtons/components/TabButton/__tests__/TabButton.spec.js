import React from 'react';
import { ctmTheme } from '@comparethemarketau/manor-themes';
import userEvent from '@testing-library/user-event';
import { render } from '../../../../../../testUtils';
import 'jest-styled-components';
import TabButton from '../TabButton.component';

describe('TabButton', () => {
  it('renders with props', () => {
    const onClick = jest.fn();
    const handleOnClick = jest.fn();
    const onRef = jest.fn();

    const { container } = render(
      <TabButton selected={false} href="#" label="test" onClick={onClick} onRef={onRef} handleOnClick={handleOnClick} />,
    );
    expect(container).toMatchSnapshot();
  });

  it('add focus visible styles ', () => {
    const onClick = jest.fn();
    const handleOnClick = jest.fn();
    const onRef = jest.fn();

    const { container } = render(
      <TabButton selected href="#" label="test" onClick={onClick} onRef={onRef} handleOnClick={handleOnClick} />,
    );

    userEvent.tab();
    const anchorTag = container.querySelector('a');
    const listElement = container.querySelector('li');

    expect(anchorTag).toHaveFocus();
    expect(listElement).toHaveStyleRule({
      'a:focus-visible': {
        outline: `${ctmTheme.colors.primary500} auto 1px`,
        outlineOffset: '-1px',
      },
    });
  });

  it('add hover styles ', () => {
    const onClick = jest.fn();
    const handleOnClick = jest.fn();
    const onRef = jest.fn();

    const { container } = render(
      <TabButton selected={false} href="#" label="test" onClick={onClick} onRef={onRef} handleOnClick={handleOnClick} />,
    );

    const listElement = container.querySelector('li');
    expect(listElement).toHaveStyleRule({
      ':hover': {
        background: `${ctmTheme.button.tertiary.backgroundHover}`,
      },
    });
  });
});
