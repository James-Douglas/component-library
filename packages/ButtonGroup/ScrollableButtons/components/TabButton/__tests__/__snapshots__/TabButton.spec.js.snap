// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabButton renders with props 1`] = `
.c1 {
  margin-top: 0;
  margin-bottom: 0;
  font-family: SourceSansPro,Arial,sans-serif;
  text-align: left;
  color: inherit;
  font-weight: 400;
  line-height: 1.4;
  font-size: 1.6rem;
}

.c0 {
  display: inline-block;
  list-style: none;
}

.c0 a:focus-visible {
  outline: #0b60b7 auto 1px;
  outline-offset: -1px;
}

.c2 {
  padding-top: 1.2rem;
  padding-right: 2.4rem;
  padding-bottom: 1.2rem;
  padding-left: 2.4rem;
  font-size: 1.4rem;
  color: #616161;
}

@media (hover:hover) {
  .c0:hover {
    background: #e7f0fb;
  }
}

<div>
  <li
    class="c0"
  >
    <a
      class="MuiTypography-root MuiLink-root MuiLink-underlineNone MuiTypography-colorPrimary"
      href="#"
      style="font-family: \\"SourceSansPro\\", \\"Helvetica\\", \\"Arial\\"; color: rgb(11, 96, 183);"
    >
      <p
        class="c1 c2"
      >
        test
      </p>
    </a>
  </li>
</div>
`;
